### PR TITLE
chore: handle logged in user migration use case

### PR DESCRIPTION
**Pull request by ami-aman**
closes: https://linear.app/customerio/issue/MOB-60/update-all-migration-methods-to-use-timestamp-and-originaltimestamp

### About the PR
This PR covers a use case scenario where a user migrates from the Journeys module to the CDP module while already logged in. This ensures the CDP module is informed about the currently logged-in user for seamless processing of events further.

### CODE REVIEW

File: Sources/Tracking/CustomerIO.swift [hunk 1]
```diff
@@ -35,15 +35,25 @@ public extension CustomerIO {
...
+        let implementation = DataPipeline.initialize(moduleConfig: DataPipelineConfigOptions.Factory.create(sdkConfig: sdkConfig))
+        let profileStore = newDiGraph.profileStore
+
+        // This code handles the scenario where a user migrates
+        // from the Journeys module to the CDP module while already logged in.
+        // This ensures the CDP module is informed about the
+        // currently logged-in user for seamless processing of events.
+        if DataPipeline.shared.analytics.userId == nil {
+            if let identifier = profileStore.identifier {
+                DataPipeline.shared.identify(identifier: identifier, body: [:])
```
Comment (mrehan27) [inline Sources/Tracking/CustomerIO.swift:48]: We should clear `profileStore.identifier` after setting it for `analytics` to make sure the code only run once.

Use case in my that could break things:

- Identify user using old SDK
- Update SDK to use `DataPipeline` -> The profile should be identified in `DataPipeline`
- Clear identify using new SDK
- Relaunch the app so SDK initialization is called again -> The will re-identify with old id in `DataPipeline` even though it was cleared before relaunch

File: Sources/Tracking/CustomerIO.swift [hunk 2]
```diff
@@ -35,15 +35,25 @@ public extension CustomerIO {
...
         var sdkConfig = SdkConfig.Factory.create(siteId: siteId, apiKey: apiKey, region: region)
+        let newDiGraph = DIGraph(sdkConfig: sdkConfig)
+        let implementation = DataPipeline.initialize(moduleConfig: DataPipelineConfigOptions.Factory.create(sdkConfig: sdkConfig))
+        let profileStore = newDiGraph.profileStore
+
+        // This code handles the scenario where a user migrates
+        // from the Journeys module to the CDP module while already logged in.
+        // This ensures the CDP module is informed about the
+        // currently logged-in user for seamless processing of events.
+        if DataPipeline.shared.analytics.userId == nil {
```
Comment (mrehan27) [inline Sources/Tracking/CustomerIO.swift:46]: It would be great if we can write a test case for it too

**Review by mrehan27 — CHANGES_REQUESTED** [review]
Requesting changes as I think we need to fix potential bug before merging

File: Sources/Tracking/CustomerIO.swift (hunk 1 above)
Comment (ami-aman) [reply Sources/Tracking/CustomerIO.swift:48]: Great catch on the use case, Rehan ! I have made the necessary updates and you can find the changes i this [commit](https://github.com/customerio/customerio-ios/pull/490/commits/4aee0591e4b9c4624a4f100484de0d7a9ca0b62e).

File: Sources/Tracking/CustomerIO.swift (hunk 2 above)
Comment (levibostian) [reply Sources/Tracking/CustomerIO.swift:46]: For readability and testability, I think it would be good to move this code into it's own encapsulated migration class.

File: Sources/Tracking/CustomerIO.swift (hunk 1 above)
Comment (levibostian) [reply Sources/Tracking/CustomerIO.swift:48]: I think a test should be written to confirm this use case is handled.

**Review by levibostian — CHANGES_REQUESTED** [review]
I would like to request automated tests before the PR gets merged.

**Comment by ami-aman** [comment]
> I would like to request automated tests before the PR gets merged.

Updated. Requested for re-review.

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift [hunk 3]
```diff
@@ -71,4 +74,25 @@ class DataPipelineMigrationAssistantTests: UnitTest {
...
+        XCTAssertNotNil(migrationAssistant.handleAlreadyIdentifiedMigratedUser())
+        XCTAssertNil(DataPipeline.shared.analytics.userId)
+    }
+
+    func test_givenUserOnCDPIdentified_expectNoUpdate() {
+        let givenIdentifier = String.random
+        DataPipeline.shared.identify(identifier: givenIdentifier)
+        XCTAssertNotNil(migrationAssistant.handleAlreadyIdentifiedMigratedUser())
+        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)
+        XCTAssertNil(profileStoreMock.identifier)
```
Comment (mrehan27) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:96]: Shouldn't it be like this?

```suggestion
        let givenIdentifier = String.random

        profileStoreMock.identifier = String.random
        DataPipeline.shared.identify(identifier: givenIdentifier)
        
        XCTAssertNotNil(migrationAssistant.handleAlreadyIdentifiedMigratedUser())
        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)
        XCTAssertNil(profileStoreMock.identifier)
```

File: Sources/Tracking/CustomerIO.swift [hunk 4]
```diff
@@ -35,16 +35,17 @@ public extension CustomerIO {
...
+        let migrationAssistant = newDiGraph.dataPipelineMigrationAssistant
 
         if let configureHandler = configureHandler {
             configureHandler(&sdkConfig)
         }
 
-        let implementation = DataPipeline.initialize(moduleConfig: DataPipelineConfigOptions.Factory.create(sdkConfig: sdkConfig))
-
+        // Handle logged-in user from Journeys to CDP
+        migrationAssistant.handleAlreadyIdentifiedMigratedUser()
```
Comment (mrehan27) [inline Sources/Tracking/CustomerIO.swift:47]: I'm okay with this, but I would have preferred making only 1 call to `migrationAssistant` here, where it could further break things inside the class and do what it wants to do. e.g.

```swift
migrationAssistant.handleMigration()
```

and inside

```swift
func handleMigration() {
    handleAlreadyIdentifiedMigratedUser()
    handleQueueBacklog()
}
```

**Review by mrehan27 — APPROVED** [review]
Non-blocking comments, looks good to me otherwise 👍🏻

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift (hunk 3 above)
Comment (ami-aman) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:96]: The method is `test_givenUserOnCDPIdentified_expectNoUpdate` i.e if CDP module has already identified a user then we do not expect the method to update anything. In your suggestion, you expect `profileStoreMock.identifier` to be `nil` which does not matter in this case. Does this help ?

File: Sources/Tracking/CustomerIO.swift (hunk 4 above)
Comment (ami-aman) [reply Sources/Tracking/CustomerIO.swift]: Thanks for the suggestion, find changes [here](https://github.com/customerio/customerio-ios/pull/490/commits/05ddf03e2b45379c67de1cac6347930e7618e6b6).

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift (hunk 3 above)
Comment (mrehan27) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:96]: I actually only wanted to make sure that `profileStore` has some identifier added to it, so added `profileStoreMock.identifier = String.random` only.

But looking at this again, I think there should never be a case where a customer is identified on CDP before `profileStore` was cleared, so we can ignore this.

File: Sources/Tracking/migration/DataPipelineMigrationAssistant.swift [hunk 5]
```diff
@@ -2,14 +2,21 @@ import CioDataPipelines
 import CioInternalCommon
 import Foundation
 
+public protocol DataPipelineMigration: AutoMockable {
+    func handleAlreadyIdentifiedMigratedUser()
+    func handleQueueBacklog()
+    func getAndProcessTask(for task: QueueTaskMetadata)
```
Comment (levibostian) [inline Sources/Tracking/migration/DataPipelineMigrationAssistant.swift:8]:
```suggestion
```

Nit-picky, but this function feels like an internal/private function only needed inside of `DataPipelineMigrationAssistant` implementation. So, not sure why we need to expose it in a protocol.

File: Sources/Tracking/migration/DataPipelineMigrationAssistant.swift [hunk 6]
```diff
@@ -20,12 +27,34 @@ class DataPipelineMigrationAssistant {
...
         self.threadUtil = threadUtil
+        self.profileStore = profileStore
+    }
+
+    func handleMigration() {
+        handleAlreadyIdentifiedMigratedUser()
+        handleQueueBacklog()
+    }
+
+    func handleAlreadyIdentifiedMigratedUser() {
```
Comment (levibostian) [inline Sources/Tracking/migration/DataPipelineMigrationAssistant.swift:45]: Besides migrating the profile ID, does our SDK also need to migrate the currently registered device token? 

If so, it can be in a separate PR.

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift [hunk 7]
```diff
@@ -11,17 +11,20 @@ class DataPipelineMigrationAssistantTests: UnitTest {
...
     private var migrationAssistant: DataPipelineMigrationAssistant { diGraph.dataPipelineMigrationAssistant }
     private var queueStorage: QueueStorage { diGraph.queueStorage }
 
     override func setUp() {
         super.setUp()
 
+        diGraph.override(value: profileStoreMock, forType: ProfileStore.self)
         diGraph.override(value: dateUtilStub, forType: DateUtil.self)
         diGraph.override(value: backgroundQueueMock, forType: Queue.self)
+        diGraph.override(value: migrationAssistantMock, forType: DataPipelineMigration.self)
```
Comment (levibostian) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:27]: I am curious why this test class is creating an instance of `migrationAssistant` via `diGraph.dataPipelineMigrationAssistant` instead of [what is commonly found in other test classes by using the class constructor](https://github.com/customerio/customerio-ios/blob/56584e39684da0b1ee4d642e43fcbaaee879d213/Tests/Common/Background%20Queue/QueueRunRequestTest.swift#L20-L27).

Comment (levibostian) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:27]: Why does `diGraph.override(value: migrationAssistantMock, forType: DataPipelineMigration.self)` exist? 

This class is testing the `DataPipelineMigration` implementation, so it shouldn't be mocked?

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift [hunk 8]
```diff
@@ -71,4 +74,25 @@ class DataPipelineMigrationAssistantTests: UnitTest {
         XCTAssertNotNil(migrationAssistant.handleQueueBacklog())
         XCTAssertEqual(backgroundQueueMock.deleteProcessedTaskCallsCount, 0)
     }
+
+    func test_givenAlreadyIdentifiedProfile_expectUpdateUserId() {
+        let givenProfileIdentifiedInJourneys = String.random
+        profileStoreMock.identifier = givenProfileIdentifiedInJourneys
+        XCTAssertNotNil(migrationAssistant.handleAlreadyIdentifiedMigratedUser())
+        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenProfileIdentifiedInJourneys)
+        XCTAssertNil(profileStoreMock.identifier)
```
Comment (levibostian) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:83]:
```suggestion
        let givenProfileIdentifiedInJourneys = String.random
        profileStoreMock.identifier = givenProfileIdentifiedInJourneys        
        XCTAssertNil(DataPipeline.shared.analytics.userId)
        
        migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        
        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenProfileIdentifiedInJourneys)
        XCTAssertNil(profileStoreMock.identifier)
```

I think this test could be improved by adding the `XCTAssertNil(DataPipeline.shared.analytics.userId)` line to make sure the state of the SDK is what we expect before and after calling `migrationAssistant.handleAlreadyIdentifiedMigratedUser()`.

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift
Comment (levibostian) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:1]: I think we need another test that calls `handleAlreadyIdentifiedMigratedUser` multiple times and we expect a migration to only occur the first time we call it and not future times that we call it.

**Review by levibostian — CHANGES_REQUESTED** [review]
I added suggestions to make the tests cover more use cases.

File: Sources/Tracking/migration/DataPipelineMigrationAssistant.swift (hunk 6 above)
Comment (ami-aman) [reply Sources/Tracking/migration/DataPipelineMigrationAssistant.swift:45]: Device token should be updated on [initialising the SDK](https://github.com/customerio/customerio-ios/blob/10168a3e6807c52404d66e96aeb49ec881f3956d/Sources/Common/CustomerIOInstance.swift#L128) from stored value in globalDataStore.

Comment (ami-aman) [reply Sources/Tracking/migration/DataPipelineMigrationAssistant.swift:45]: If this make sense to you then please feel free to resolve the conversation.

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift (hunk 7 above)
Comment (ami-aman) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:27]: I am removing `migrationAssistantMock` as it is not in use and was added as a part of a dev test.

File: Tests/Common/Extensions/DateExtensionTests.swift [hunk 9]
```diff
@@ -59,4 +59,18 @@ class DateExtensionTest: UnitTest {
...
+    // MARK: formatToIso8601WithMilliseconds
+
+    func test_givenDate_formatToIso8601WithMilliseconds_expectString() {
+        let date = dateUtilStub.now
+        let formattedDate = date.string(format: .iso8601WithMilliseconds)
+
+        let formatter = ISO8601DateFormatter()
+        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
+        formatter.timeZone = TimeZone(identifier: "UTC")
+        let expectedDate = formatter.string(from: date)
```
Comment (levibostian) [inline Tests/Common/Extensions/DateExtensionTests.swift:72]: This test might accurately test the use case, but unless I fully understand this highlighted block of code, I cannot be certain. 

Also, is this highlighted block of code has a bug, then we cannot be sure if the code under test is working or not. 

As a suggestion, I suggest using a hard-coded string as the expected value. Something like: 

```swift
    func test_givenDate_formatToIso8601WithMilliseconds_expectString() {
        let givenDate = Date(...)
        let expectedString = "2023-12-04 ..."

        let actualString = givenDate.string(format: .iso8601WithMilliseconds)

        XCTAssertEqual(expectedString, actualString)
     }
```

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift
Comment (levibostian) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:1]: Thanks for writing a test for this. 

Looking at the test function body, it doesn't test the use case that I was referring to. 

The use case I had in mind was: 
* A profile is identified in Journeys SDK
* The SDK gets initialized, so the CDP migration occurs. 
* The app gets killed. 
* The SDK get initialized the next time the app opens. 
* Because the SDK has been migrated to use the CDP SDK, we expect the CDP migration to not happen again. 

I was expecting code similar to: 
```swift
        let givenIdentifier = String.random
        profileStore.identifier = givenIdentifier 

        XCTAssertNil(DataPipeline.shared.analytics.userId)
        let didMigrationRun = migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        AssertTrue(didMigrationRun)
        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)

        // The next time the SDK gets initialized and the migration runs again: 
        let didMigrationRunAgain = migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        AssertFalse(didMigrationRunAgain)
    }
```

Comment (ami-aman) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:1]: @levibostian `handleAlreadyIdentifiedMigratedUser` does not return anything. I have written a similar test case. The userId gets updated only once. If it has already been set then we do not expect any change. Request you to recheck the test case and let me know if there's a miss ?

File: Tests/Common/Extensions/DateExtensionTests.swift (hunk 9 above)
Comment (ami-aman) [reply Tests/Common/Extensions/DateExtensionTests.swift]: Updated https://github.com/customerio/customerio-ios/pull/490/commits/fa796e28a0e2e8150c301e4afce7af98a3f389c8

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift [hunk 10]
```diff
@@ -59,4 +64,45 @@ class DataPipelineMigrationAssistantTests: UnitTest {
...
+
+        // Update the user identifier and re-call handleAlreadyIdentifiedMigratedUser
+        // to ensure the user does not undergo the migration process again
+        // after being identified on the CDP
+        let updatedIdentifier = String.random
+        profileStoreMock.identifier = updatedIdentifier
+        XCTAssertNotNil(migrationAssistant.handleAlreadyIdentifiedMigratedUser())
+        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)
+        XCTAssertEqual(profileStoreMock.identifier, updatedIdentifier)
+    }
```
Comment (levibostian) [inline Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:107]: I think that this test function is trying to test [this use case](https://github.com/customerio/customerio-ios/pull/490/files#r1476464247) suggested in code review. 

If that is true, then here is a suggestion I have for this test body. 

```suggestion
    func test_migrateProfileId_expectMigrationCodeRunOnce() {
        // profile previously identified in SDK, before CDP migration
        let givenIdentifier = String.random
        profileStoreMock.identifier = updatedIdentifier
        XCTAssertNil(DataPipeline.shared.analytics.userId)
        
        migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        
        // CDP migration is performed for the first time in the SDK. 
        migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        // Check that the migration was successful:
        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)

        // Update the user identifier and re-call handleAlreadyIdentifiedMigratedUser
        // to ensure the user does not undergo the migration process again
        // after being identified on the CDP
        let updatedIdentifier = String.random
        profileStoreMock.identifier = updatedIdentifier
        
        migrationAssistant.handleAlreadyIdentifiedMigratedUser()
        
        // We expect the CDP profile ID is the same value from the 1st migration done. 
        XCTAssertEqual(DataPipeline.shared.analytics.userId, givenIdentifier)
    }
```

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift
Comment (levibostian) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:1]: Left a suggestion: https://github.com/customerio/customerio-ios/pull/490/files#r1479932261

File: Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift (hunk 10 above)
Comment (ami-aman) [reply Tests/Tracking/migration/DataPipelineMigrationAssistantTests.swift:107]: Thanks for the suggestion. The suggested code breaks, hence I am fixing it and making the update

**Review by levibostian — APPROVED** [review]
